File: solver.py
import json
import time
from selenium import webdriver
from selenium.webdriver.firefox.options import Options


def merge_data(current_data, new_data):
    for k, v in new_data.items():
        if isinstance(v, dict):
            if k not in current_data:
                current_data[k] = v
            else:
                merge_data(current_data[k], v)
        else:
            current_data[k] = v


def get_pairs(data):
    data_length = len(data)
    pair_tuples = []

    for i, j in enumerate(data):
        for k in range(i + 1, data_length):
            pair_tuples.append((j["text"], data[k]["text"]))

    return pair_tuples


def main():
    options = Options()
    options.headless = True
    driver = webdriver.Firefox(options=options)

    driver.get("https://neal.fun/infinite-craft")

    with open("data.json", "r") as file:
        current_data = json.load(file)["elements"]

    all_pairs = get_pairs(current_data)
    for i in range(len(all_pairs)):
        url = f"https://neal.fun/api/infinite-craft/pair?first={all_pairs[i][0]}&second={all_pairs[i][1]}"

        print(all_pairs[i])

        response_text = driver.execute_script(
            f"""
            var xhr = new XMLHttpRequest();
            xhr.open("GET", "{url}", false);  // Synchronous request
            xhr.send();
            return xhr.responseText;
        """
        )

        print(response_text)

        time.sleep(0.3)

    driver.quit()


if __name__ == "__main__":
    main()

File: solver.js
function allStorage() {

    var values = [],
        keys = Object.keys(localStorage),
        i = keys.length;

    while (i--) {
        values.push(localStorage.getItem(keys[i]));
    }

    return values;
}

function getPairs(foundElements) {
    let elements = [...foundElements]
    let pairs = [];
    elements.push({
        text: "Earth",
        emoji: "🌍",
        discovered: false
    })
    elements.push({
        text: "Fire",
        emoji: "🔥",
        discovered: false
    })
    elements.push({
        text: "Water",
        emoji: "💧",
        discovered: false
    })
    elements.push({
        text: "Wind",
        emoji: "🌬️",
        discovered: false
    })
    for (let i = 0; i < elements.length; i++) {
        for (let j = i + 1; j < elements.length; j++) {
            pairs.push([elements[i].text, elements[j].text]);
        }
    }
    return pairs;
}

function hasObjectWithSameFields(arr, obj) {
    for (let i = 0; i < arr.length; i++) {
        const arrayObj = arr[i];

        if (Object.keys(arrayObj).length === Object.keys(obj).length) {
            let sameFields = true;
            for (const key in obj) {
                if (obj.hasOwnProperty(key) && !arrayObj.hasOwnProperty(key)) {
                    sameFields = false;
                    break;
                }
            }
            if (sameFields) {
                return true;
            }
        }
    }
    return false;
}

var sleepSetTimeout_ctrl;

function sleep(ms) {
    clearInterval(sleepSetTimeout_ctrl);
    return new Promise(resolve => sleepSetTimeout_ctrl = setTimeout(resolve, ms));
}

async function test() {

    var elementIndex = -1;
    // figure out the index in allstorage that has element
    for (let i = 0; i < allStorage().length; i++) {
        if (allStorage()[i].includes("elements")) {
            elementIndex = i;
        }
    }

    var elements = JSON.parse(allStorage()[elementIndex]).elements;
    for (let pair of getPairs(elements)) {
        const xhr = new XMLHttpRequest();
        xhr.open("GET", `https://neal.fun/api/infinite-craft/pair?first=${pair[0]}&second=${pair[1]}`);
        xhr.send();
        xhr.responseType = "json";
        xhr.onload = () => {
            if (xhr.readyState == 4 && xhr.status == 200) {
                const data = xhr.response;
                console.log(data);

                const exists = elements.some(element =>
                    element.text === data.result &&
                    element.emoji === data.emoji &&
                    element.discovered === data.isNew
                );

                if (!exists && data.text != 'Nothing' && data.emoji != '') {
                    elements.push({ text: data.result, emoji: data.emoji, discovered: data.isNew });
                    localStorage.setItem("infinite-craft-data", JSON.stringify({ elements: elements }));
                }
            }
        };
        await sleep(300);
    }
}


test();
